File: sorting/sorting.js


//--- heapsort: in-place implementation with linear-time heap construction

function trickleDown(data, pos, n) {
	var left = 2*pos+1;
	if (left >= n) {
		return;
	}
	var right = 2*pos+2;
	if (right >= n) {
		if (data[pos] < data[left]) {
			var t = data[pos];
			data[pos] = data[left];
			data[left] = t;
		}
		return;
	}
	if (data[pos] < data[left]) {
		if (data[left] < data[right]) {
			var t = data[pos];
			data[pos] = data[right];
			data[right] = t;
			trickleDown(data, right, n);
		} else {
			var t = data[pos];
			data[pos] = data[left];
			data[left] = t;
			trickleDown(data, left, n);
		}
	} else if(data[pos] < data[right]) {
		var t = data[pos];
		data[pos] = data[right];
		data[right] = t;
		trickleDown(data, right, n);
	}
}

function heapsort(data) {
	// build a max-heap
	var n = data.length;
	for (var i = Math.floor((n-2)/2); i != -1; --i) {
		trickleDown(data, i, n);
	}

	// successively remove the max item and swap it to the end
	for (var i = n-1; i != -1; --i) {
		var t = data[i];
		data[i] = data[0];
		data[0] = t;
		trickleDown(data, 0, i);
	}
}

function testHeapsort(data) {
	timeFunction('ran heapsort', function () {
		heapsort(data);
	});
	if (!isSorted(data)) {
		print('error: data is not sorted');
	}
}


//--- mergesort: recursive implementation with two arrays

function mergesort(data, buffer, left, right) {
	if (left+1 == right) {
		return;
	}
	var mid = Math.floor((left+right)/2);
	mergesort(data, buffer, left, mid);
	mergesort(data, buffer, mid, right);
	var leftPos = left, rightPos = mid,
		bufferPos = left;
	while (true) {
		if (data[leftPos] <= data[rightPos]) {
			buffer[bufferPos++] = data[leftPos++];
			if (leftPos == mid) {
				while (rightPos != right) {
					buffer[bufferPos++] = data[rightPos++];
				}
				break;
			}
		} else {
			buffer[bufferPos++] = data[rightPos++];
			if (rightPos == right) {
				while (leftPos != mid) {
					buffer[bufferPos++] = data[leftPos++];
				}
				break;
			}
		}
	}
	for (var i = left; i != right; ++i) {
		data[i] = buffer[i];
	}
}

function testMergesort(data) {
	timeFunction('ran mergesort', function () {
		var buffer = new Array(data.length);
		mergesort(data, buffer, 0, data.length);
	});
	if (!isSorted(data)) {
		print('error: data is not sorted');
	}
}


//--- quicksort: in-place implementation with median-of-three pivot selection

function quicksort(data, left, right) {
	if (left+1 >= right) {
		return;
	}
	var ai = left, bi = Math.floor((left+right)/2), ci = right-1,
		a = data[ai], b = data[bi], c = data[ci],
		pos;
    if (a < b) {
        if (c < a) {
            pos = ai;
		} else if (c < b) {
			pos = ci;
		} else {
			pos = bi;
		}
	} else {
        if (c < b) {
            pos = bi;
		} else if (c < a) {
			pos = ci;
		} else {
			pos = ai;
		}
	}
    var pivot = data[pos];
	data[pos] = data[right-1];
	var tail = left;
	for (var i = left; i != right-1; ++i) {
		if (data[i] < pivot) {
			var t = data[tail];
			data[tail] = data[i];
			data[i] = t;
			++tail;
		}
	}
	data[right-1] = data[tail];
	data[tail] = pivot;
    quicksort(data, left, tail);
    quicksort(data, tail+1, right);
}

function testQuicksort(data) {
	timeFunction('ran quicksort', function () {
		quicksort(data, 0, data.length);
	});
	if (!isSorted(data)) {
		print('error: data is not sorted');
	}
}

function quicksortCustom(data, left, right, compare) {
	if (left+1 >= right) {
		return;
	}
	var ai = left, bi = Math.floor((left+right)/2), ci = right-1,
		a = data[ai], b = data[bi], c = data[ci],
		pos;
    if (compare(a, b) < 0) {
        if (compare(c, a) < 0) {
            pos = ai;
		} else if (compare(c, b) < 0) {
			pos = ci;
		} else {
			pos = bi;
		}
	} else {
		if (compare(c, b) < 0) {
            pos = bi;
		} else if (compare(c, a) < 0) {
			pos = ci;
		} else {
			pos = ai;
		}
	}
    var pivot = data[pos];
	data[pos] = data[right-1];
	var tail = left;
	for (var i = left; i != right-1; ++i) {
		if (compare(data[i], pivot) < 0) {
			var t = data[tail];
			data[tail] = data[i];
			data[i] = t;
			++tail;
		}
	}
	data[right-1] = data[tail];
	data[tail] = pivot;
    quicksortCustom(data, left, tail, compare);
    quicksortCustom(data, tail+1, right, compare);
}

function testQuicksortCustom(data, compare) {
	timeFunction('ran quicksort', function () {
		quicksortCustom(data, 0, data.length, compare);
	});
	if (!isSortedCustom(data, compare)) {
		print('error: data is not sorted');
	}
}


//--- native sort

function testNativeSort(data, compare) {
	if (compare) {
		var fun = function () {
			data.sort(compare);
		}
	} else {
		var fun = function () {
			data.sort();
		}
	}
	timeFunction('ran native sort', fun);
	if (!isSorted(data)) {
		print('error: data is not sorted');
	}
}


//--- random permutation

function shuffle(data) {
	// constants for a linear congruential generator
	var m = Math.pow(2, 32),
		a = 1664525,
		c = 1013904223,
		randomValue = 42;  // seed
	for (var i = data.length-1; i != 0; --i) {
		randomValue = (a*randomValue + c) % m;
		var j = randomValue % (i+1);
		if (j != i) {
			var t = data[i];
			data[i] = data[j];
			data[j] = t;
		}
	}
}


//--- helper functions

function timeFunction(description, fun) {
	var startTime = performance.now();
	fun();
	var elapsed = (performance.now() - startTime)/1000;
	print('    '+description+'  '+formatF(elapsed, 3)+' s');
}

function isSorted(data) {
	var n = data.length;
	for (var i = 1; i != n; ++i) {
		if (data[i-1] > data[i]) {
			return false;
		}
	}
	return true;
}

function isSortedCustom(data, compare) {
	var n = data.length;
	for (var i = 1; i != n; ++i) {
		if (compare(data[i-1], data[i]) > 0) {
			return false;
		}
	}
	return true;
}

function displayStudents(students) {
	var parts = [], n = students.length;
	for (var i = 0; i < n; ++i) {
		parts.push(students[i].grade+' '+students[i].name);
	}
	print(parts.join(' '));
}

function formatF(value, digits) {
	var whole = Math.floor(value),
		decimal = ''+Math.round(Math.pow(10, digits)*(value-whole));
	decimal = decimal+'00000000000000000000'.slice(0, digits-decimal.length);
	return whole+"."+decimal;
}


function main() {
	var n = parseInt(readline());
	if (!(n >= 1)) {
		print('error: must specify the length of the array');
		return;
	}


	//--- integer sorting
	print('integers:');

	// initialize an array with 0, ..., n-1
	var sortedIntegers;
	timeFunction('made ordered integer array', function () {
		sortedIntegers = new Array(n);
		for (var i = 0; i < n; ++i) {
			sortedIntegers[i] = i;
		}
	});

	// permute randomly
	var shuffledIntegers = sortedIntegers.slice();
	timeFunction('shuffled', function () {
		shuffle(shuffledIntegers);
	});

	// sort integer arrays
	testHeapsort(shuffledIntegers.slice());
	testMergesort(shuffledIntegers.slice());
	testQuicksort(shuffledIntegers.slice());
	testNativeSort(shuffledIntegers.slice(), function (a, b) {
		return a-b;
	});


	//--- string sorting
	print('strings:');

	// make strings: 'a', 'b', ..., 'z', 'aa', 'ab', ..., 'az', 'ba', ...
	var sortedStrings;
	timeFunction('made ordered string array', function () {
		sortedStrings = new Array(n);
		var next_char = {}, zCode = 'z'.charCodeAt(0);
		for (var code = 'a'.charCodeAt(0); code != zCode; ++code) {
			next_char[String.fromCharCode(code)] = String.fromCharCode(code+1);
		}
		var chars = ['a'], numChars = 1;
		for (var i = 0; i != n; ++i) {
			sortedStrings[i] = chars.join('');
			var pos = numChars-1;
			while (true) {
				if (chars[pos] != 'z') {
					chars[pos] = next_char[chars[pos]];
					break;
				}
				pos -= 1;
				if (pos == -1) {
					chars.unshift('a');
					numChars += 1;
					break;
				}
			}
			if (pos != numChars-1) {
				for (var j = pos+1; j <= numChars-1; ++j) {
					chars[j] = 'a';
				}
			}
		}
	});

	// permute randomly
	var shuffledStrings = sortedStrings.slice();
	timeFunction('shuffled', function () {
		shuffle(shuffledStrings);
	});

	// sort string arrays
	testQuicksort(shuffledStrings.slice());
	testNativeSort(shuffledStrings.slice());

	// object sorting
	print('objects:');

	// make student objects based on the sorted strings
	var sortedStudents = new Array(n);
	timeFunction('made student objects', function () {
		for (var i = 0; i < n; ++i) {
			sortedStudents[i] = { name: sortedStrings[i], grade: i%10 };
		}
	});

	// permute randomly
	var shuffledStudents = sortedStudents.slice();
	timeFunction('shuffled', function () {
		shuffle(shuffledStudents);
	});

	// sort objects
	function compareStudents(a, b) {
		if (a.grade != b.grade) {
			return a.grade - b.grade;
		} else if (a.name < b.name) {
			return -1;
		} else if (a.name > b.name) {
			return 1;
		}
		return 0;
	}
	testQuicksortCustom(shuffledStudents.slice(), compareStudents);
	testNativeSort(shuffledStudents.slice(), compareStudents);
}

main();

File: sorting/Sorting.java
import java.lang.*;
import java.util.*;
import java.io.*;

class Student implements Comparable<Student> {
	String name;
	int grade;
	Student(String name, int grade) {
		this.name = name;
		this.grade = grade;
	}
	public String toString() {
		return grade+" "+name;
	}
	public int compareTo(Student other) {
		if (this.grade != other.grade) {
			return this.grade - other.grade;
		}
		return this.name.compareTo(other.name);
	}
}

public class Sorting {

	//--- heapsort: in-place implementation with linear-time heap construction

	static <T extends Comparable<T>> void trickleDown(ArrayList<T> data,
			int pos, int n) {
		int left = 2*pos+1;
		if (left >= n) {
			return;
		}
		int right = 2*pos+2;
		if (right >= n) {
			if (data.get(pos).compareTo(data.get(left)) < 0) {
				Collections.swap(data, pos, left);
			}
			return;
		}
		if (data.get(pos).compareTo(data.get(left)) < 0) {
			if (data.get(left).compareTo(data.get(right)) < 0) {
				Collections.swap(data, pos, right);
				trickleDown(data, right, n);
			} else {
				Collections.swap(data, pos, left);
				trickleDown(data, left, n);
			}
		} else if(data.get(pos).compareTo(data.get(right)) < 0) {
			Collections.swap(data, pos, right);
			trickleDown(data, right, n);
		}
	}

	static <T extends Comparable<T>> void heapsort(ArrayList<T> data) {
		// build a max-heap
		int n = data.size();
		for (int i = (n-2)/2; i != -1; --i) {
			trickleDown(data, i, n);
		}

		// successively remove the max item and swap it to the end
		for (int i = n-1; i != -1; --i) {
			Collections.swap(data, i, 0);
			trickleDown(data, 0, i);
		}
	}

	static <T extends Comparable<T>> void testHeapsort(ArrayList<T> data) {
		double startTime = System.currentTimeMillis();
		heapsort(data);
		double elapsed = (System.currentTimeMillis() - startTime)/1000;
		System.out.println(String.format(
					"    ran heapsort  %.3f", elapsed));
		if (!isSorted(data)) {
			System.out.println("error: data is not sorted");
		}
	}


	//--- mergesort: recursive implementation with two arrays

	static <T extends Comparable<T>> void mergesort(ArrayList<T> data, ArrayList<T> buffer,
			int left, int right) {
		if (left+1 == right) {
			return;
		}
		int mid = (left+right)/2;
		mergesort(data, buffer, left, mid);
		mergesort(data, buffer, mid, right);
		int leftPos = left, rightPos = mid,
			bufferPos = left;
		while (true) {
			if (data.get(leftPos).compareTo(data.get(rightPos)) <= 0) {
				buffer.set(bufferPos++, data.get(leftPos++));
				if (leftPos == mid) {
					while (rightPos != right) {
						buffer.set(bufferPos++, data.get(rightPos++));
					}
					break;
				}
			} else {
				buffer.set(bufferPos++, data.get(rightPos++));
				if (rightPos == right) {
					while (leftPos != mid) {
						buffer.set(bufferPos++, data.get(leftPos++));
					}
					break;
				}
			}
		}
		for (int i = left; i != right; ++i) {
			data.set(i, buffer.get(i));
		}
	}

	static <T extends Comparable<T>> void testMergesort(ArrayList<T> data) {
		double startTime = System.currentTimeMillis();
		ArrayList<T> buffer = new ArrayList<T>(data);
		mergesort(data, buffer, 0, data.size());
		double elapsed = (System.currentTimeMillis() - startTime)/1000;
		System.out.println(String.format(
					"    ran mergesort  %.3f", elapsed));
		if (!isSorted(data)) {
			System.out.println("error: data is not sorted");
		}
	}


	//--- quicksort: in-place implementation with median-of-three pivot

	static <T extends Comparable<T>> void quicksort(ArrayList<T> data, int left, int right) {
		if (left+1 >= right) {
			return;
		}
		int ai = left, bi = (left+right)/2, ci = right-1, pos;
		T a = data.get(ai), b = data.get(bi), c = data.get(ci);
		if (a.compareTo(b) < 0) {
			if (c.compareTo(a) < 0) {
				pos = ai;
			} else if (c.compareTo(b) < 0) {
				pos = ci;
			} else {
				pos = bi;
			}
		} else {
			if (c.compareTo(b) < 0) {
				pos = bi;
			} else if (c.compareTo(a) < 0) {
				pos = ci;
			} else {
				pos = ai;
			}
		}
		T pivot = data.get(pos);
		Collections.swap(data, right-1, pos);
		int tail = left;
		for (int i = left; i != right-1; ++i) {
			if (data.get(i).compareTo(pivot) < 0) {
				Collections.swap(data, tail, i);
				++tail;
			}
		}
		Collections.swap(data, right-1, tail);
		quicksort(data, left, tail);
		quicksort(data, tail+1, right);
	}

	static <T extends Comparable<T>> void testQuicksort(ArrayList<T> data) {
		double startTime = System.currentTimeMillis();
		quicksort(data, 0, data.size());
		double elapsed = (System.currentTimeMillis() - startTime)/1000;
		System.out.println(String.format(
					"    ran quicksort  %.3f", elapsed));
		if (!isSorted(data)) {
			System.out.println("error: data is not sorted");
		}
	}


	//--- native sort

	static <T extends Comparable<T>> void testNativeSort(ArrayList<T> data) {
		double startTime = System.currentTimeMillis();
		Collections.sort(data);
		double elapsed = (System.currentTimeMillis() - startTime)/1000;
		System.out.println(String.format(
					"    ran native sort  %.3f", elapsed));
		if (!isSorted(data)) {
			System.out.println("error: data is not sorted");
		}
	}


	//--- random permutation

	static <T extends Comparable<T>> void shuffle(ArrayList<T> data) {
		// constants for a linear congruential generator
		long m = (1L << 32),
			 a = 1664525,
			 c = 1013904223,
			 randomValue = 42;  // seed
		for (int i = data.size()-1; i != 0; --i) {
			randomValue = (a*randomValue + c) % m;
			int j = (int)(randomValue % (long)(i+1));
			if (j != i) {
				Collections.swap(data, i, j);
			}
		}
	}


	//--- helper methods

	static <T extends Comparable<T>> boolean isSorted(ArrayList<T> data) {
		int n = data.size();
		for (int i = 1; i != n; ++i) {
			if (data.get(i-1).compareTo(data.get(i)) > 0) {
				return false;
			}
		}
		return true;
	}

	static <T extends Comparable<T>> void display(ArrayList<T> data) {
		System.out.print(data.get(0));
		for (int i = 1; i != data.size(); ++i) {
			System.out.print(" "+data.get(i));
		}
		System.out.println();
	}


	public static void main(String args[]) {
		if (args.length != 1) {
			System.out.println("error: must specify the length of the array");
			return;
		}
		int n = Integer.valueOf(args[0]);

		//--- integer sorting
		System.out.println("integers:");

		// initialize an array with 0, ..., n-1
		double startTime = System.currentTimeMillis();
		ArrayList<Integer> sortedIntegers = new ArrayList<Integer>();
		for (int i = 0; i < n; ++i) {
			sortedIntegers.add(i);
		}
		double elapsed = (System.currentTimeMillis() - startTime)/1000;
		System.out.println(String.format(
				"    made ordered integer array  %.3f", elapsed));

		// permute randomly
		ArrayList<Integer> shuffledIntegers =
				new ArrayList<Integer>(sortedIntegers);
		startTime = System.currentTimeMillis();
		shuffle(shuffledIntegers);
		elapsed = (System.currentTimeMillis() - startTime)/1000;
		System.out.println(String.format("    shuffled  %.3f", elapsed));

		// sort integer arrays
		testHeapsort(new ArrayList<Integer>(shuffledIntegers));
		testMergesort(new ArrayList<Integer>(shuffledIntegers));
		testQuicksort(new ArrayList<Integer>(shuffledIntegers));
		testNativeSort(new ArrayList<Integer>(shuffledIntegers));


		//--- string sorting
		System.out.println("strings:");

		// make strings: 'a', 'b', ..., 'z', 'aa', 'ab', ..., 'az', 'ba', ...
		startTime = System.currentTimeMillis();

		ArrayList<String> sortedStrings = new ArrayList<String>();
		StringBuffer buf = new StringBuffer("a");
		int numChars = 1;
		for (int i = 0; i != n; ++i) {
			sortedStrings.add(buf.toString());
			int pos = numChars-1;
			while (true) {
				if (buf.charAt(pos) != 'z') {
					buf.setCharAt(pos, (char)(buf.charAt(pos)+1));
					break;
				}
				pos -= 1;
				if (pos == -1) {
					buf.insert(0, 'a');
					++numChars;
					break;
				}
			}
			if (pos != numChars-1) {
				for (int j = pos+1; j <= numChars-1; ++j) {
					buf.setCharAt(j, 'a');
				}
			}
		}
		elapsed = (System.currentTimeMillis() - startTime)/1000;
		System.out.println(String.format(
				"    made ordered string array  %.3f", elapsed));

		// permute randomly
		ArrayList<String> shuffledStrings =
				new ArrayList<String>(sortedStrings);
		startTime = System.currentTimeMillis();
		shuffle(shuffledStrings);
		elapsed = (System.currentTimeMillis() - startTime)/1000;
		System.out.println(String.format("    shuffled  %.3f", elapsed));

		// sort string arrays
		testQuicksort(new ArrayList<String>(shuffledStrings));
		testNativeSort(new ArrayList<String>(shuffledStrings));


		//--- object sorting
		System.out.println("objects:");

		// make student objects based on the sorted strings
		startTime = System.currentTimeMillis();
		ArrayList<Student> sortedStudents = new ArrayList<Student>();
		for (int i = 0; i < n; ++i) {
			sortedStudents.add(new Student(sortedStrings.get(i), i%10));
		}
		elapsed = (System.currentTimeMillis() - startTime)/1000;
		System.out.println(String.format(
				"    made student objects  %.3f", elapsed));

		// permute randomly
		ArrayList<Student> shuffledStudents =
				new ArrayList<Student>(sortedStudents);
		startTime = System.currentTimeMillis();
		shuffle(shuffledStudents);
		elapsed = (System.currentTimeMillis() - startTime)/1000;
		System.out.println(String.format("    shuffled  %.3f", elapsed));

		// sort objects
		testQuicksort(new ArrayList<Student>(shuffledStudents));
		testNativeSort(new ArrayList<Student>(shuffledStudents));
	}
}
